### PR TITLE
Add PureJSON renderer

### LINKED ISSUE #693: Special HTML chars encoded in JSON respones?
I've noticed that, in responses from `gin.Context.JSON`,  special html characters get escaped, while other unicode characters are treated normally.

Here's a handler to reproduce this:

```
func Hello(c *gin.Context) {
    c.JSON(200, map[string]string{
        "message": "<foo> zażółć gęślą jaźń",
    })
}
```

```
$ curl localhost:3000/api/v1/
{"message":"\u003cfoo\u003e zażółć gęślą jaźń"}
```

### PR DESCRIPTION
**Pull request by megapctr**
Closes #693

### CODE REVIEW

**Comment by megapctr** [comment]
Looks like it's not compiling :laughing:
The error is:
```
@PureJSON $ ~/gin > go test
# _/home/filip/gin
./context.go:427: undefined: render.PureJSON
FAIL    _/home/filip/gin [build failed]
```
could someone help me out? The code looks good to me.

**Comment by cristaloleg** [comment]
@megapctr Maybe you should add ```PureJSON interface{}``` here: https://github.com/rikuayanokozy/gin/blob/743cdd1a39fefa28e4e6f86d81a02b82512b433d/context.go#L499

**Comment by megapctr** [comment]
@cristaloleg the problem is that this feature is available in go 1.7 only. I don't know how to handle that for earlier versions. Any tips?

**Comment by megapctr** [comment]
I'm referring to `encoder.SetEscapeHTML` specifically

**Comment by robvdl** [comment]
Sorry for being a bit blunt, but the way you could be dealing with the Go 1.7 issue is by first releasing version 1.0 of Gin, it's about time to put a line in the sand and doing a 1.0 release (Gin to me seems a bit "stuck" to me as 1.0 never comes out). Then any Gin releases after 1.0 making them for Go 1.7 only.  Then users of Go 1.6 and below can still use the 1.0 release if they want.

**Comment by megapctr** [comment]
I see. It's ok for me to wait until gin is ready for the next step. Besides, this PR isn't any crucial feature at all. Sending a json works fine either way. With PureJSON you cut a few bytes off from the response size, which may be helpful in some rare cases

**Comment by appleboy** [comment]
@megapctr Could you fix the conflicts?

**Comment by appleboy** [comment]
@megapctr conflicts.

File: render/render_test.go [hunk 1]
```diff
@@ -42,13 +42,30 @@ func TestRenderMsgPack(t *testing.T) {
...
 
 	err := (JSON{data}).Render(w)
 
 	assert.NoError(t, err)
 	assert.Equal(t, "{\"foo\":\"bar\"}", w.Body.String())
+	assert.Equal(
+		t,
+		w.Body.String(),
+		"{\"foo\":\"bar\",\"html\":\"\\u003cb\\u003e\"}\n")
+	assert.Equal(t, w.Header().Get("Content-Type"), "application/json; charset=utf-8")
```
Comment (thinkerou) [inline render/render_test.go]: should be `assert.Equal(t, expect, actual)`

**Comment by thinkerou** [comment]
@megapctr can you help to add example to `readme`? thanks!

**Comment by thinkerou** [comment]
compile error: `render.PureJSON does not implement render.Render (missing WriteContentType method)`

**Comment by megapctr** [comment]
I have trouble running tests locally so I'll rely on travis

**Comment by megapctr** [comment]
@thinkerou you mean the main readme? I think it would be identical to the existing example, except it would use `c.PureJSON` instead of `c.JSON`.

What should I do to support go 1.6?

**Comment by thinkerou** [comment]
@megapctr add https://github.com/gin-gonic/gin#jsonp after. Now still support go1.6.

**Comment by appleboy** [comment]
build fail

```
go test -v -covermode=count -coverprofile=coverage.out
# github.com/gin-gonic/gin/render
render/json.go:59: encoder.SetEscapeHTML undefined (type *"encoding/json".Encoder has no field or method SetEscapeHTML)
FAIL	github.com/gin-gonic/gin [build failed]
make: *** [test] Error 2
```

**Comment by megapctr** [comment]
What should I do about go 1.6?

**Comment by megapctr** [comment]
Travis hates me today 😢 Apart from that is everything looking ok?

File: context_17.go [hunk 2]
```diff
@@ -0,0 +1,17 @@
+// Copyright 2014 Manu Martinez-Almeida.  All rights reserved.
+// Use of this source code is governed by a MIT style
+// license that can be found in the LICENSE file.
```
Comment (appleboy) [inline context_17.go]:
```
// Copyright 2018 Gin Core Team.  All rights reserved.
// Use of this source code is governed by a MIT style
// license that can be found in the LICENSE file.
```

File: context_17_test.go [hunk 3]
```diff
@@ -0,0 +1,26 @@
+// Copyright 2014 Manu Martinez-Almeida.  All rights reserved.
+// Use of this source code is governed by a MIT style
+// license that can be found in the LICENSE file.
+
```
Comment (appleboy) [inline context_17_test.go]:
```
// Copyright 2018 Gin Core Team.  All rights reserved.
// Use of this source code is governed by a MIT style
// license that can be found in the LICENSE file.
```

File: render/json_17.go [hunk 4]
```diff
@@ -0,0 +1,28 @@
+// Copyright 2014 Manu Martinez-Almeida.  All rights reserved.
+// Use of this source code is governed by a MIT style
+// license that can be found in the LICENSE file.
```
Comment (appleboy) [inline render/json_17.go]:
```
// Copyright 2018 Gin Core Team.  All rights reserved.
// Use of this source code is governed by a MIT style
// license that can be found in the LICENSE file.
```

File: render/render_17_test.go [hunk 5]
```diff
@@ -0,0 +1,26 @@
+// Copyright 2014 Manu Martinez-Almeida.  All rights reserved.
+// Use of this source code is governed by a MIT style
+// license that can be found in the LICENSE file.
+
```
Comment (appleboy) [inline render/render_17_test.go]:
```
// Copyright 2018 Gin Core Team.  All rights reserved.
// Use of this source code is governed by a MIT style
// license that can be found in the LICENSE file.
```

**Comment by appleboy** [comment]
@thinkerou I don't know why `make vet` only can't work on `go latest` version. maybe we can ignore the error from Travis.

**Comment by thinkerou** [comment]
Oh....@appleboy Makefile not update lead to the error, because the pull request commit to `develop` branch. so recommit to master branch or change it(how change?).

**Comment by thinkerou** [comment]
@megapctr thanks a lot, firstly, can you fix some code according to my review result, secondly, please recommit the pull request to master branch, because it will error which pointed above. thanks again.

**Comment by bjm88** [comment]
Hi, specifically for context.PureJSON - in latest gin version dont see this as option...did I miss something or this never got into release from master?  How to render json without escaping?

**Pull request merged**